File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.java
class Solution {
    public boolean canPartition(int[] nums) {
        int totalSum = Arrays.stream(nums).sum();
        if (totalSum % 2 != 0) return false;
        int target = totalSum / 2;
        int[][] cache = new int[nums.length][target + 1];


        return dfs(0, nums, target, cache);

    }

    private boolean dfs(int i, int[] nums, int target, int[][] cache) {
        if (target == 0) return true;
        if (target < 0 || i == nums.length) return false;

        if (cache[i][target] != 0) {
            return cache[i][target] == 1;
        }


        cache[i][target] = (dfs(i + 1, nums, target, cache) || dfs(i + 1, nums, target - nums[i], cache)) ? 1 : -1;
        
        return cache[i][target] == 1;
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-3.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans=[]
        nums.sort()
        n = len(nums)
        for i in range(0,n):
            if i>0 and nums[i]==nums[i-1]:
                continue
            target=-1*nums[i]
            left = i+1
            right = n-1
            while right>left:
                sum = nums[left]+nums[right]
                if sum>target:
                    right-=1
                elif sum<target:
                    left+=1
                else:
                    ans.append([nums[i],nums[left],nums[right]])
                    while left<right and nums[left]==nums[left+1]:
                        left+=1
                    left+=1
                    while right>left and nums[right]==nums[right-1]:
                        right-=1
                    right-=1
        return ans

